File: Q2683_「ロング・ロング・ストリング」問題/answer.lua
-- Try CodeIQ
-- author: Leonardone @ NEETSDKASU

function f(n)
	return 1 + math.floor(n * math.log10(n))
end

g = function (m)
		h = function (b, t)
				if b > t then return -1 end
				x = math.floor((b + t) / 2)
				y = f(x)
				if y > m then
					return h(b, x - 1)
				elseif y < m then
					return h(x + 1, t)
				else
					return x
				end
			end
		return h(2, 10 ^ 10)
	end

m = tonumber(io.read())
ans = g(m)
print(ans)

File: Q2683_「ロング・ロング・ストリング」問題/answer.f
! Try CodeIQ
! author: Leonardone @ NEETSDKASU

PROGRAM CodeIQ
    REAL(8) :: m, b, t, x, y 
    INTEGER(8) :: ans

    READ(*,*) m
    b = 2.0
    t = 10000000000.0
    ans = -1
    DO WHILE (b <= t)
        x = FLOOR((b + t) / 2.0, 8)
        y = 1 + FLOOR(x * LOG10(x), 8) ! 間違っていたので修正
        IF (y < m) THEN
            b = x + 1.0
        ELSE IF (y > m) THEN
            t = x - 1.0
        ELSE
            ans = FLOOR(x, 8)
            EXIT
        END IF
    END DO
    PRINT '(I0)', ans
    
    STOP
END PROGRAM CodeIQ